gdb/rust-lang.c

```diff
@@ -1183,7 +1415,139 @@ rust_evaluate_subexp (struct type *expect_type, struct expression *exp,
+                                     value_address(lhs), lhs);
+
+
+          if (disr.field_no < 0) {
+            error (_("Could not determine variant of enum value"));
+          }
+
+          cleanup = make_cleanup (xfree, disr.name);
+
+          variant_type = TYPE_FIELD_TYPE(type, disr.field_no);
+          nfields = TYPE_NFIELDS (variant_type);
+
+          // note that there is the extra discriminant field as well,
+          // so the actual field number is field_number + 1
+          if (field_number >= nfields - 1 || field_number < 0) {
+            // unsure how to call error with a cleanup
+            error(_("Cannot access field %d of variant %s, \
+there are only %d fields"),
+                  field_number, disr.name, nfields - 1);
+          }
+
+          if (!rust_tuple_variant_type_p (variant_type)) {
+            error(_("Variant %s is not a tuple variant"),
+                  disr.name);
+          }
+
+          result = value_primitive_field(lhs, 0,
+                                         field_number + 1, variant_type);
+          do_cleanups (cleanup);
+          break;
+        } else if (TYPE_CODE (type) == TYPE_CODE_STRUCT) {
+          /* Tuples and tuple structs */
+          nfields = TYPE_NFIELDS(type);
+
+          if (field_number >= nfields || field_number < 0) {
+            error(_("Cannot access field %d of %s, there are only %d fields"),
+                  field_number, TYPE_TAG_NAME (type), nfields);
+          }
+
+          /* tuples are tuple structs too */
+          if (!rust_tuple_struct_type_p (type)) {
+            error(_("Attempting to access anonymous field %d of %s, which is \
+not a tuple, tuple struct, or tuple-like variant"),
+                  field_number, disr.name);
+          }
+
+          result = value_primitive_field(lhs, 0, field_number, type);
+          break;
+        }
+        error(_("Anonymous field access is only allowed on tuples, \
+tuple structs, and tuple-like enum variants"));
+      }
+      break;
+    case STRUCTOP_STRUCT:
+      {
+        struct value* lhs;
+        struct type *type;
+        int tem, pc;
+
+        pc = (*pos)++;
+        tem = longest_to_int (exp->elts[pc + 1].longconst);
+        (*pos) += 3 + BYTES_TO_EXP_ELEM (tem + 1);
+        lhs = evaluate_subexp (NULL_TYPE, exp, pos, noside);
+
+        type = value_type (lhs);
+
+        if (TYPE_CODE (type) == TYPE_CODE_UNION) {
+
+          int i;
+          struct disr_info disr;
+          struct cleanup* cleanup;
+          struct type* variant_type;
+          char* field_name;
 
+          field_name = &exp->elts[pc + 2].string;
+
+          disr = rust_get_disr_info (type, value_contents(lhs),
+                                     value_embedded_offset(lhs),
+                                     value_address(lhs), lhs);
+
+
+          if (disr.field_no < 0) {
+            error (_("Could not determine variant of enum value"));
```

Comment:
If all the callers do this, it's fine to make `rust_get_disr_info` call `error`.  Either way is ok with me though.